### PR TITLE
Support JSON as on-disk format for histograms

### PR DESCRIPTION
**Pull request by idoqo**
This adds support for JSON as an on-disk format for histograms as described in [MDEV-21130](https://jira.mariadb.org/browse/MDEV-21130).

At the moment, JSON histograms can now be enabled with: `set histogram_type=JSON;` and running `ANALYZE TABLE` as usual.

Some improvements over the existing histogram format is also tracked at [MDEV-26125](https://jira.mariadb.org/browse/MDEV-26125).

Signed-off-by: Michael Okoko <okokomichaels@outlook.com>

### CODE REVIEW

**Comment by cvicentiu** [comment]
Hi @idoqo!

Good progress. What your commit is missing is a test case showcasing the functionality.

If you haven't figured out how to write test cases yet, you can read this:

https://mariadb.org/get-involved/getting-started-for-developers/writing-good-test-cases-mariadb-server/

File: sql/sql_statistics.h [hunk 1]
```diff
@@ -179,6 +182,7 @@ class Histogram
     case SINGLE_PREC_HB:
       return (uint) (((uint8 *) values)[i]);
     case DOUBLE_PREC_HB:
+    case JSON:
```
Comment (cvicentiu) [inline sql/sql_statistics.h:185]: So currently the code behaves identical for JSON type and DOUBLE precision. That's fine for a first iteration.

File: sql/sql_statistics.h [hunk 2]
```diff
@@ -154,6 +155,7 @@ class Histogram
     case SINGLE_PREC_HB:
       return ((uint) (1 << 8) - 1);
     case DOUBLE_PREC_HB:
+    case JSON:
       return ((uint) (1 << 16) - 1);
```
Comment (spetrunia) [inline sql/sql_statistics.h:159]: It's ok for now, but in the future this code should be removed as the concept of `prec_factor` doesn't apply to Histograms that are stored as JSON.

**Comment by spetrunia** [comment]
Ok, this is good for Milestone-1

**Comment by cvicentiu** [comment]
@idoqo, @spetrunia is now working on getting Json histograms into 10.8. as such the PR can be considered merged. I am closing this PR given this. 

f there are fixes to add, please sync with @spetrunia on which branch to base them on and open a new PR for those.

**Comment by cvicentiu** [comment]
@idoqo And just to clarify what I mean:

The work you have done has made it as a preview release in 10.7.0, but it didn't make the cut into 10.7.1, due to regressions we have identified.

 @spetrunia is tracking the progress of the task here: https://jira.mariadb.org/browse/MDEV-21130 with different subtasks for current outstanding parts. You are more than welcome to contribute to those.
 
 The work is now happening on this branch:
 https://github.com/MariaDB/server/tree/preview-10.7-MDEV-26519-json-histograms

**Comment by idoqo** [comment]
Got it, thanks for the heads up!

**Pull request closed without merging**